onnx/shape_inference/implementation.h

```diff
@@ -140,9 +141,10 @@ struct InferenceContextImpl : public InferenceContext {
       const std::unordered_map<std::string, TypeProto*>& valueTypesByName,
       const std::unordered_map<std::string, const TensorProto*>& inputDataByName,
       const std::unordered_map<std::string, const SparseTensorProto*>& inputSparseDataByName,
+      const ShapeInferenceOptions& options,
```

Comment:
Would it better that setting a default "options" here? (by default all false for example) For those users who already relied on this function, they don't need to change their code if they don't care strict_mode.

Reply:
@gramalingam @jcwchen you both commented in the same direction which is to relax requirement for providing ShapeInferenceOptions. In this sprit, I can add  ShapeInferenceOptions as a member of GraphInferencerImpl and remove both getShapeInferenceOptions method and the "const ShapeInferenceOptions& options" argument from doInferencing method. The drawback is that, for example in onnxruntim, it create a different GraphInferencerImpl which can totally ignore providing ShapeInferenceOptions.

Reply:
> The drawback is that, for example in onnxruntim, it create a different GraphInferencerImpl which can totally ignore providing ShapeInferenceOptions.

It seems fine to me that any runtime can choose whether to provide ShapeInferenceOptions or not. Also I think it's good to maintain backward compatibility here by having a default ShapeInferenceOption. Perhaps we can add a TODO saying in the future ONNX should require every shape inference related functions to honor ShapeInferenceOptions.

Reply:
I have relaxed the API - not to require to provide ShapeInferenceOption.